crates/block-producer/src/mempool/state_dag.rs

```diff
@@ -0,0 +1,442 @@
+        for tx in selected {
+            // SAFETY: Selected txs come from the transaction pool and are unique.
+            tx_nodes.push(self.nodes.txs.remove(&tx).unwrap());
+        }
+
+        let batch_id =
+            BatchId::from_transactions(tx_nodes.iter().map(|tx| tx.0.raw_proven_transaction()));
+        let node = ProposedBatchNode(tx_nodes);
+        let batch = node.0.iter().map(|tx| Arc::clone(&tx.0)).collect();
+
+        self.nodes.proposed_batches.insert(batch_id, node);
+        // FIXME: missing safety
+        self.insert_into_state_unchecked(NodeId::ProposedBatch(batch_id));
+
+        Some((batch_id, batch))
+    }
+
+    pub fn revert_proposed_batch(&mut self, batch: BatchId) {
+        // Due to the distributed nature of the system, its possible that a proposed batch was
+        // already proven, or already reverted. This guards against this eventuality.
+        if !self.nodes.proposed_batches.contains_key(&batch) {
+            return;
+        }
+
+        // FIXME: missing safety
+        self.revert_subtree_unchecked(NodeId::ProposedBatch(batch));
+    }
+
+    pub fn submit_batch_proof(&mut self, proof: Arc<ProvenBatch>) {
+        // Due to the distributed nature of the system, its possible that a proposed batch was
+        // already proven, or already reverted. This guards against this eventuality.
+        let Some(ProposedBatchNode(txs)) = self.nodes.proposed_batches.remove(&proof.id()) else {
+            return;
+        };
+
+        // Propagate the node ID change internally.
+        let batch_id = proof.id();
+        let node = ProvenBatchNode { txs, inner: proof };
+        self.nodes.proven_batches.insert(batch_id, node);
+
+        // FIXME: missing safety
+        self.insert_into_state_unchecked(NodeId::ProvenBatch(batch_id));
+    }
+
+    pub fn propose_block(
+        &mut self,
+        mut budget: BlockBudget,
+    ) -> (BlockNumber, Vec<Arc<ProvenBatch>>) {
+        assert!(
+            self.nodes.proposed_block.is_none(),
+            "block {} is already in progress",
+            self.nodes.proposed_block.as_ref().unwrap().0
+        );
+
+        let mut selected = Vec::new();
+
+        'outer: loop {
+            'next_candidate: for candidate in
+                self.nodes.proven_batches.values().map(|node| &node.inner)
+            {
+                if selected.contains(&candidate.id()) {
+                    continue;
+                }
+
+                let parents = self.parents(NodeId::ProvenBatch(candidate.id()));
+                for parent in parents {
+                    match parent {
+                        NodeId::ProposedBlock(block) => panic!(
+                            "Block selection encountered an existing proposed block ID {block} while navigating the state DAG"
+                        ),
+                        NodeId::ProvenBatch(batch) if selected.contains(&batch) => {},
+                        NodeId::CommittedBlock(_) => {},
```

Comment:
If a proven batch has a committed block as a parent, doesn't that mean it is already committed, or if not, what are the circumstances under which this could happen? If it is already commited I we would have to `continue 'next_candidate`, right?

In any case, I think a one line comment above each non-obvious case might be helpful.